Adds tutorial: Space Shuttle Reentry Trajectory

ferrolho opened the pull request:
Hi! :slightly_smiling_face: I would like to propose this notebook to be added under `JuMPTutorials.jl/notebook/modelling`. It is about finding a trajectory for the reentry of the Space Shuttle. You can check the notebook online [here](https://nbviewer.jupyter.org/github/ferrolho/JuMPTutorials.jl/blob/master/notebook/modelling/space_shuttle_reentry_trajectory.ipynb). Thanks!

ferrolho committed:
Adds tutorial: Space Shuttle Reentry Trajectory

matbesancon commented:
Hi, thanks for the contribution. One thing to note to contribute, it is preferable to write the tutorial in `script/folder/tutorial.jl` using a syntax similar to the others, it is using https://github.com/JunoLab/Weave.jl to convert this script to notebooks and other formats

ferrolho commented:
Sorry, I totally missed [this page](https://jump.dev/JuMPTutorials.jl/dev/new/). I will check what needs to be done to make this properly.

matbesancon commented:
Thanks! Let us know if you have any questions.

On Sun, Feb 7, 2021 at 7:41 PM Henrique Ferrolho <notifications@github.com>
wrote:

> Sorry, I totally missed this page
> <https://jump.dev/JuMPTutorials.jl/dev/new/>. I will check what needs to
> be done to make this properly.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/jump-dev/JuMPTutorials.jl/pull/97#issuecomment-774729981>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB2FDMTLFXNYXQMEMNIG2WDS53NHJANCNFSM4XHWNO2A>
> .
>


-- 
Mathieu Besançon

ferrolho commented:
Hi again! :slightly_smiling_face: I think I have now added my contribution in the proper way:
1. Created my notebook at `script/modelling/space_shuttle_reentry_trajectory.jl`
2. Generated the test and notebook files programmatically with
    ```julia
    using JuMPTutorials
    JuMPTutorials.weave_file("modelling", "space_shuttle_reentry_trajectory.jl")
    ```
3. Added myself to the contributors list in `docs/src/index.md`
4. Added a link to the notebook in `README.md`

ferrolho commented:
I noticed [Weave.jl](https://github.com/JunoLab/Weave.jl) was not on the dependencies but I think it should be (see terminal log below), so I added it to the dependencies list. On a similar note, my tutorial also relies on [Interpolations.jl](https://github.com/JuliaMath/Interpolations.jl), so I added that too. Let me know if this is a problem, though.
```
julia> using JuMPTutorials
[ Info: Precompiling JuMPTutorials [767a2c38-803a-11e9-191e-47b580bc80c9]
ERROR: LoadError: ArgumentError: Package JuMPTutorials does not have Weave in its dependencies:
- If you have JuMPTutorials checked out for development and have
  added Weave as a dependency but haven't updated your primary
  environment's manifest file, try `Pkg.resolve()`.
- Otherwise you may need to report an issue with JuMPTutorials
Stacktrace:
 [1] require(::Module, ::Symbol) at ./loading.jl:906
 [2] include(::Function, ::Module, ::String) at ./Base.jl:380
 [3] include(::Module, ::String) at ./Base.jl:368
 [4] top-level scope at none:2
 [5] eval at ./boot.jl:331 [inlined]
 [6] eval(::Expr) at ./client.jl:467
 [7] top-level scope at ./none:3
in expression starting at /home/henrique/git/JuMPTutorials.jl/src/JuMPTutorials.jl:3
ERROR: Failed to precompile JuMPTutorials [767a2c38-803a-11e9-191e-47b580bc80c9] to /home/henrique/.julia/compiled/v1.5/JuMPTutorials/S3rQy_TftkL.ji.
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] compilecache(::Base.PkgId, ::String) at ./loading.jl:1305
 [3] _require(::Base.PkgId) at ./loading.jl:1030
 [4] require(::Base.PkgId) at ./loading.jl:928
 [5] require(::Module, ::Symbol) at ./loading.jl:923

(JuMPTutorials) pkg> add Weave
  Resolving package versions...
  Installed YAML ─────── v0.4.6
  Installed Highlights ─ v0.4.5
  Installed Weave ────── v0.10.6
Updating `~/git/JuMPTutorials.jl/Project.toml`
  [44d3d7a6] + Weave v0.10.6
Updating `~/git/JuMPTutorials.jl/Manifest.toml`
  [eafb193a] + Highlights v0.4.5
  [ffc61752] + Mustache v1.0.10
  [44d3d7a6] + Weave v0.10.6
  [ddb6d928] + YAML v0.4.6

julia> using JuMPTutorials
[ Info: Precompiling JuMPTutorials [767a2c38-803a-11e9-191e-47b580bc80c9]
WARNING: Method definition latex(IO, Markdown.LaTeX) in module Markdown at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/Markdown/src/IPython/IPython.jl:28 overwritten in module WeaveMarkdown at /home/henrique/.julia/packages/Weave/BarbQ/src/WeaveMarkdown/markdown.jl:10.
  ** incremental compilation may be fatally broken for this module **

WARNING: Method definition latex(IO, Markdown.LaTeX) in module Markdown at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/Markdown/src/IPython/IPython.jl:28 overwritten in module WeaveMarkdown at /home/henrique/.julia/packages/Weave/BarbQ/src/WeaveMarkdown/markdown.jl:10.
  ** incremental compilation may be fatally broken for this module **
```

ferrolho commented:
(Sorry for the chain of comments.)

Finally, every time I ran
```julia
JuMPTutorials.weave_file("modelling", "space_shuttle_reentry_trajectory.jl")
```
I got this error
```
ERROR: IOError: could not spawn `jupyter nbconvert --ExecutePreprocessor.timeout=-1 --to notebook --execute /home/henrique/git/JuMPTutorials.jl/notebook/modelling/space_shuttle_reentry_trajectory.ipynb --allow-errors --output /home/henrique/git/JuMPTutorials.jl/notebook/modelling/space_shuttle_reentry_trajectory.ipynb`: no such file or directory (ENOENT)
Stacktrace:
 [1] _spawn_primitive(::String, ::Cmd, ::Array{Any,1}) at ./process.jl:99
 [2] #585 at ./process.jl:112 [inlined]
 [3] setup_stdios(::Base.var"#585#586"{Cmd}, ::Array{Any,1}) at ./process.jl:196
 [4] _spawn at ./process.jl:111 [inlined]
 [5] open(::Cmd, ::Base.DevNull; write::Bool, read::Bool) at ./process.jl:370
 [6] open at ./process.jl:361 [inlined]
 [7] open(::Cmd, ::String, ::Base.DevNull) at ./process.jl:341
 [8] read(::Cmd) at ./process.jl:410
 [9] read(::Cmd, ::Type{String}) at ./process.jl:421
 [10] notebook(::String; out_path::String, timeout::Int64, nbconvert_options::String, jupyter_path::String) at /home/henrique/.julia/packages/Weave/BarbQ/src/Weave.jl:289
 [11] weave_file(::String, ::String) at /home/henrique/git/JuMPTutorials.jl/src/JuMPTutorials.jl:32
 [12] top-level scope at REPL[14]:1
```
but everything seemed to actually work fine... :man_shrugging:

matbesancon commented:
That's a problem on my end, I should remove the files in src, doing it now

frapac commented on test/modelling/space_shuttle_reentry_trajectory.jl:
```diff
@@ -0,0 +1,179 @@
+
+user_options_ipopt = (
+    "mu_strategy" => "monotone",
+    "linear_solver" => "ma27",
```
By setting this option, I am wondering if Ipopt fallbacks to Mumps if not compiled with `ma27`. I would maybe add a comment to explain that.

ferrolho replied:
Something like?
```
"linear_solver" => "ma27",  # Ipopt may fallback to Mumps, if not compiled with ma27.
```

frapac replied:
In that line. I would maybe precise that MA27 is generally faster than MUMPS (a lot of people using Ipopt do not know that we could change the underlying linear solver)

frapac commented on script/modelling/space_shuttle_reentry_trajectory.jl:
```diff
@@ -0,0 +1,281 @@
+#' $$
+
+#' The aerodynamic and atmospheric forces on the vehicle are
+#' specified by the following quantities (English units):
```
I was not aware that this example is using English units :-p

ferrolho replied:
It is. :slightly_smiling_face: I implemented the problem as close as possible to the book version.

ferrolho replied:
Actually, I copied almost word-by-word the problem description from the book to the start of the notebook. (I.e., the markdown content of the current version.)

![image](https://user-images.githubusercontent.com/4166093/107356308-89a18700-6ac8-11eb-8f49-3da2fc74e6ed.png)

I hope this is not an issue. I do cite the book immediately at the start, so proper credit is given.

frapac replied:
I don't think this is an issue, as you stated, the book is correctly cited :-)

frapac commented on script/modelling/space_shuttle_reentry_trajectory.jl:
```diff
@@ -0,0 +1,281 @@
+#+
+
+# Global variables
+const w  = 203000  # weight (lb)
```
I would maybe specify `w` as a flow`:
```
const w = 203000.0
```

ferrolho replied:
You mean renaming `w` as `flow`? I don't have any preference, so I am happy to change any variable names. However, do note this will deviate from the original notation in the book.

frapac replied:
My bad: I meant `float` instead of `flow`, but my fingers slipped on my keyboard...

frapac commented on script/modelling/space_shuttle_reentry_trajectory.jl:
```diff
@@ -0,0 +1,281 @@
+
+user_options_ipopt = (
+    "mu_strategy" => "monotone",
+    "linear_solver" => "ma27",
```
Same comment as below about ma27. I am wondering whether we should include a comment there

frapac commented on script/modelling/space_shuttle_reentry_trajectory.jl:
```diff
@@ -0,0 +1,281 @@
+                          ψ[1:n]                # azimuth (rad)
+    deg2rad(-90) ≤        α[1:n] ≤ deg2rad(90)  # angle of attack (rad)
+    deg2rad(-89) ≤        β[1:n] ≤ deg2rad( 1)  # banck angle (rad)
+    #        0.5 ≤       Δt[1:n] ≤ 1.5          # time step (sec)
```
Why is this line commented out?

ferrolho replied:
Good question. There are two lines in the source code for specifying time:
```julia
    #        0.5 ≤       Δt[1:n] ≤ 1.5          # time step (sec)
                         Δt[1:n] == 1.0         # time step (sec)
```
The first one (currently commented) allows some wiggle room for the solver to adjust the time step size between pairs of mesh points. This is neat because it allows the solver to figure out which parts of the flight require more dense discretization than others. (Remember, the number of discretised points is fixed, and this example does not implement mesh refinement.) However, this makes the problem more complex to solve, and therefore leads to a longer computation time.

The second line (the one not commented) fixes the duration of every time step to exactly 1.0 seconds. This allows the problem to be solved faster. The solver is still able to find a solution to the problem because I "cheated": I know that the close-to-optimal total duration of the flight is ~33 minutes, so I set the number of knots to `n = 2009` for the fixed time step to be set to 1.0 seconds. Does that make sense?

frapac commented on script/modelling/space_shuttle_reentry_trajectory.jl:
```diff
@@ -0,0 +1,281 @@
+    @NLconstraint(model, γ[j] == γ[i] + Δt[i] * δγ[i])
+    @NLconstraint(model, ψ[j] == ψ[i] + Δt[i] * δψ[i])
+
+    # # Trapezoidal integration
```
Do you think we could include a new option to switch between rectangular and trapezoidal integration?

ferrolho replied:
Of course! What do you have in mind? Would something like this work?
```julia
# (...)

integration_rule = "rectangular"  # Possible values: rectangular, trapezoidal.

# (...)

# System dynamics
for j in 2:n
    i = j - 1  # index of previous knot

    if integration_rule == "rectangular"
        # Rectangular integration
        @NLconstraint(model, h[j] == h[i] + Δt[i] * δh[i])
        @NLconstraint(model, ϕ[j] == ϕ[i] + Δt[i] * δϕ[i])
        @NLconstraint(model, θ[j] == θ[i] + Δt[i] * δθ[i])
        @NLconstraint(model, v[j] == v[i] + Δt[i] * δv[i])
        @NLconstraint(model, γ[j] == γ[i] + Δt[i] * δγ[i])
        @NLconstraint(model, ψ[j] == ψ[i] + Δt[i] * δψ[i])
    elseif integration_rule == "trapezoidal"
        # Trapezoidal integration
        @NLconstraint(model, h[j] == h[i] + 0.5 * Δt[i] * (δh[j] + δh[i]))
        @NLconstraint(model, ϕ[j] == ϕ[i] + 0.5 * Δt[i] * (δϕ[j] + δϕ[i]))
        @NLconstraint(model, θ[j] == θ[i] + 0.5 * Δt[i] * (δθ[j] + δθ[i]))
        @NLconstraint(model, v[j] == v[i] + 0.5 * Δt[i] * (δv[j] + δv[i]))
        @NLconstraint(model, γ[j] == γ[i] + 0.5 * Δt[i] * (δγ[j] + δγ[i]))
        @NLconstraint(model, ψ[j] == ψ[i] + 0.5 * Δt[i] * (δψ[j] + δψ[i]))
    else
        @error "Unexpected integration rule '$(integration_rule)'"
    end
end

# (...)
```

frapac replied:
That would be perfect

frapac reviewed (commented):
Nice tutorial! I think it is useful, as the model is non-trivial and could be quite illustrative for non-linear programming. If I had to make a course on non-linear optimization, I may take this tutorial as an example.

matbesancon commented:
@ferrolho I made a PR to fix the weave issue #111, once this is merged you can rebase on master

matbesancon commented on Project.toml:
```diff
@@ -24,6 +25,7 @@ Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
 SCS = "c946c3f1-0d1f-5ce8-9dea-7daa1f7e2d13"
 Statistics = "10745b16-79ce-11e8-11f9-7d13ad32a3b2"
 Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
+Weave = "44d3d7a6-8a23-5bf8-98c5-b353f8df5ec9"
```
```suggestion
```

Not required anymore

ferrolho committed:
[WIP] Tutorial: Space Shuttle Reentry Trajectory

What's new?
- Adds Interpolations.jl to project dependencies
- Adds new notebook under `script/modelling`
- Adds test and notebook files generated with `JuMPTutorials.weave_file("modelling", "space_shuttle_reentry_trajectory.jl")`
- Adds Henrique to contributors

ferrolho committed:
Adds variable for integration scheme selection

ferrolho committed:
Makes `w` a Float

ferrolho committed:
Adds note about linear solver user option

ferrolho committed:
Final adjustments to space shuttle tutorial

Brings user variables up to the first code cell and makes them const. These include the initial and final conditions, as well as the number of mesh points and integration rule to be used for the optimisation.

ferrolho committed:
Adds outputs of notebook after running all cells

ferrolho commented:
Hi! Thank you all for your very good suggestions. I believe I have addressed all the comments. The updated notebook file can be seen [here](https://nbviewer.jupyter.org/github/ferrolho/JuMPTutorials.jl/blob/master/notebook/modelling/space_shuttle_reentry_trajectory.ipynb) (same link as before).

odow commented:
Want to move this to the main JuMP repository? You can find the tutorials here: https://github.com/jump-dev/JuMP.jl/tree/master/docs/src/tutorials

ferrolho commented:
> Want to move this to the main JuMP repository?

Yes! Here it is: https://github.com/jump-dev/JuMP.jl/pull/2496 :slightly_smiling_face:

odow commented:
Great! Closing in favor of the JuMP tutorial.
